### PR TITLE
Finally fixed flakey `TimeoutError` false negatives in CI

### PR DESCRIPTION
**Pull request by tombh**
## Description
Related to the recent flakey timeout errors we've been seeing on Windows CI runs, I discovered that when a test randomly raises a `TimeoutError` the test server could not be stopped, thus preventing the entire test suite from exiting, and ultimately preventing us from seeing the underlying problem.

This commit does not resolve any reasons why a test might occasionally exceed its timeout, but it does allow the test suite to exit, which then allows us to identify problematic tests.

However, the current WIP fix does not work on Windows! But at least it's a step in the right direction as CI is actually exiting and giving us an error now, namely that Windows more often than not does not raise `TimeoutError` even when set to `1e-50` seconds.

Any ideas welcome...

## Code review checklist (for code reviewer to complete)

- [x] Pull request represents a single change (i.e. not fixing disparate/unrelated things in a single PR)
- [x] Title summarizes what is changing
- [x] Commit messages are meaningful (see [this][commit messages] for details)
- [x] Tests have been included and/or updated, as appropriate
- [x] Docstrings have been included and/or updated, as appropriate
- [ ] Standalone docs have been updated accordingly
- [x] [CONTRIBUTORS.md][contributors] was updated, as appropriate
- [ ] Changelog has been updated, as needed (see [CHANGELOG.md][changelog])

### CODE REVIEW

**Comment by tombh** [comment]
I think I've got a fix! I've run it 5 times in a row, and not a single flakey timeout on any OS 🤓

This was quite a journey. The short of it is that there is a bug where subsequent initialisations of test suite's `ClientServer` hang. That I haven't got to the bottom of yet, but what I have done is wrap it in a retry decorator, just so we can isolate the issue and fix it when priorities allow. Here's the fix hack:
```python
# This is in conftest.py, nothing to do with the actual production source code
@retry_stalled_init_fix_hack()
def initialize(self):
    response = self.client.lsp.send_request(
        INITIALIZE,
        InitializeParams(
            process_id=12345,
            root_uri="file://",
            capabilities=ClientCapabilities()
        ),
    ).result(timeout=1)
    assert "capabilities" in response

```

There is a longer story in that this was particularly hard to debug because it turns out there were 2 distinct ways of running the test `ClientServer`, both of which could indeed be running at the same time. The first way was for the tests in `tests/test_*` and the second way was with the classes of `tests/lsp/test_*`. The `lsp/test_*` method was even more troublesome because it wasn't refactored to a single place. So all in all it was both hard to debug and hard to fix until I refactored everything to use the same isolated approach. This explains why so many files are affected by this PR. I tried to keep non-functional formatting changes in separate commits, but the main commit "fix: Unified main and lsp test-server usage" is still just huge and I suspect quite hard to review. So here's the basic gist. Before we the `lsp/*` tests looked like this:

```python
class TestCallHierarchy(unittest.TestCase):
    @classmethod
    def setUpClass(cls):
        cls.client_server = ClientServer()
        cls.client, cls.server = cls.client_server

        @cls.server.feature(
            TEXT_DOCUMENT_CALL_HIERARCHY_PREPARE,
            CallHierarchyOptions(),
        )
        def f(params: CallHierarchyPrepareParams) -> Optional[List[CallHierarchyItem]]:
        ...
        cls.client_server.start()

    @classmethod
    def tearDownClass(cls):
        cls.client_server.stop()

    def test_capabilities(self):
        capabilities = self.server.server_capabilities
        ...
```

Now they follow the same structure as the `tests/test_` tests:
```python
class ConfiguredLS(ClientServer):
    def __init__(self):
        super().__init__()

        @self.server.feature(
            TEXT_DOCUMENT_CALL_HIERARCHY_PREPARE,
            CallHierarchyOptions(),
        )
        def f1(
            params: CallHierarchyPrepareParams
        ) -> Optional[List[CallHierarchyItem]]:
        ...

@ConfiguredLS.decorate()
def test_capabilities(client_server):
    _, server = client_server
    capabilities = server.server_capabilities
    assert capabilities.call_hierarchy_provider
    ...
```

You can see that there is no setup/teardown in the classes anymore. Every single test now goes through the same isolated, debuggable function:
```python
@pytest.fixture(autouse=True)
def client_server(request):
    if hasattr(request, 'param'):
        # `lsp/*` tests
        LS = request.param
        LS()
        client_server = LS()
    else:
        # `tests/test_` tests
        client_server = ClientServer()
        setup_ls_features(client_server.server)

    client_server.start()
    client, server = client_server

    yield client, server

    client_server.stop()
```

So that's the gist of the biggest change that needs to be considered in the review.

As an aside, it occurred to me a few times that it'd be good to use https://github.com/psf/black to reduce the amount of formatting discrepancies and thus reduce the amount of diff noise in commits. But it's not essential for now.

**Review by renatav — APPROVED** [review]
I just have two questions, but since the tests are passing (which is really exciting), I'm approving the PR. I did not take a detailed look at each of the changes, just some examples of how the tests were refactored. Since they are being run and are passing, no reason to doubt that they are still correct.
Would be nice to run black. One problem that we faced with that was that it was harder to use git blame following the formatting since the most recent change ended up being "Black code refactoring" more often than not.

File: tests/conftest.py [hunk 1]
```diff
@@ -61,71 +50,76 @@ def __init__(self):
...
-
-    def _stop(self):
-        shutdown_response = (
-            self.client
-            .lsp.send_request(pygls.lsp.methods.SHUTDOWN)
-            .result(timeout=CALL_TIMEOUT)
-        )
+        shutdown_response = self.client.lsp.send_request(
+            pygls.lsp.methods.SHUTDOWN
+        ).result()
```
Comment (alcarney) [inline tests/conftest.py:84]: Should there be a timeout here?

File: tests/conftest.py [hunk 2]
```diff
@@ -61,71 +50,76 @@ def __init__(self):
...
-@pytest.fixture(scope='session', autouse=True)
-def client_server():
-    """ A fixture to setup a client/server """
-    client_server = ClientServer()
-    setup_ls_features(client_server.server)
+@pytest.fixture(autouse=True)
+def client_server(request):
+    if hasattr(request, 'param'):
+        LS = request.param
+        LS()
```
[the remark this answers is not in the document]
Comment (renatav) [reply tests/conftest.py]: I do like line 115 here, `request.param` is quite unintuitive. Byt then don't see the point of `LS()` with the next line being  `client_server = LS()`. I wanted to say why do we instantiate `LS` twice, but I doubt that that makes a difference anyway

File: tests/conftest.py [hunk 3]
```diff
@@ -43,16 +41,7 @@
...
 
-def pytest_sessionfinish(session, exitstatus):
-    """ whole test run finishes. """
-    sleep(CALL_TIMEOUT)
-    for cs in ClientServer._to_stop:
-       cs._stop()
-
-
 class ClientServer:
-    _to_stop: List['ClientServer'] = []
```
[the remark this answers is not in the document]
Comment (renatav) [reply tests/conftest.py:54]: That makes sense.

**Comment by renatav** [comment]
Looks good, unless @alcarney's comment needs to be addressed

File: tests/conftest.py (hunk 1 above)
Comment (alcarney) [reply tests/conftest.py:84]: I see, makes sense!

**Review by alcarney — APPROVED** [review]

**Pull request merged**